묵찌빠 [STEP 1] newJunsung, JaeHyeok

newJunsung opened the pull request:
**리뷰어:** @forestjae

## 파일구조
- RockPaperScissors
  + main.swift
  + RockPaperScissor.swift
  + RPCError.swift
----
## 고민했던 점
### 가위바위보 상성관계
초기에 만들었을 때, 모든 경우의 수를 체크해서 조건문으로 구현했지만,
STEP 2에서 묵찌빠의 턴과 승리조건을 만들 때에도, `if`로 모든 경우의 수를 만들어야 한다는 문제가 있을 것이라고 판단했습니다.
따라서 이 게임의 경우, 낼 수있는 경우가 3가지(가위, 바위, 보)밖에 없기 때문에, 케이스로 나누기 위해 `enum`을 만들었습니다.
그리고 가위바위보는 각각 1, 2, 3으로 받기 때문에, 이것을 enum으로 변환해주기 위한 `Int` 구조체를 상속했고,
가위, 바위, 보를 상성관계를 재사용가능하게 하기 위해, `Comparable` 프로토콜을 채택했습니다.

- **초기 버전**

main.swift
```swift
if player == 1 && computer == 1 {
        print("비겼습니다.")
    } else if player == 1 && computer == 2 {
        print("졌습니다.")
    } else {
        print("이겼습니다.")
    }
    
    
    if player == 2 && computer == 1 {
        print("이겼습니다.")
    } else if player == 2 && computer == 2 {
        print("비겼습니다.")
    } else {
        print("졌습니다.")
    }
    
    if player == 3 && computer == 1 {
        print("졌습니다.")
    } else if player == 3 && computer == 2 {
        print("이겼습니다.")
    } else {
        print("비겼습니다.")
    }
```
- **Comparable 프로토콜을 활용하여 만든 enum**

RockPaperScissors.swift
```swift
import Foundation

enum RockPaperScissors: Int, Comparable {
    case scissors = 1
    case rock = 2
    case paper = 3
    
    static func < (lhs: RockPaperScissors, rhs: RockPaperScissors) -> Bool {
        if lhs == .paper && rhs == .scissors {
            return true
        }
        if lhs == .rock && rhs == .paper {
            return true
        }
        if lhs == .scissors && rhs == .rock {
            return true
        }
        return false
    }
}
```

main.swift
```swift
if playerHand == computerHand {
        print("비겼습니다!")
        return true
    }else if computerHand < playerHand {
        print("이겼습니다!")
        return false
    }else{
        print("졌습니다!")
        return false
    }
```

### 예외처리
사용자로부터 잘못된 입력을 받았을 때, 언래핑하는 과정에서 오류가 있을 경우마다 `return`이나 `continue`를 통해 오류를 처리했지만,
이런 방법보다 예외처리를 활용하여 오류를 처리하는게 좋을 것이라고 생각했습니다.
(나중에 다른 오류가 발생해도, Error에 경우를 추가하면 되기 때문에 유지보수가 편하고 코드도 간결해지기 때문)
그래서 `enum`을 만들고 `Error`프로토콜을 채택했습니다.
- **초기 버전**
 
main.swift
```swift
guard let computer = (1...3).randomElement() else {
        print("오류가 발생했습니다.")
        return false
}

guard let playerHand = RockPaperScissors(rawValue: player), let computerHand = RockPaperScissors(rawValue: computer) else {
        print("오류가 발생했습니다.")
        return false
}
```

- **Error 프로토콜을 활용하여 만든 RPCError**

RPCError.swift
```swift
import Foundation

enum RPCError: Error {
    case invalidInputError
}
```
main.swift
```swift
func battle(player: Int) throws -> Bool {
    guard let computer = (1...3).randomElement() else {
        throw RPCError.invalidInputError
    }
    
    guard let playerHand = RockPaperScissors(rawValue: player), let computerHand = RockPaperScissors(rawValue: computer) else {
        throw RPCError.invalidInputError
    }
// ...
}
```

newJunsung committed:
docs: .gitignore 문서 수정

newJunsung committed:
feat: 가위바위보 상성관계 함수 추가

silverjaehyeok committed:
refactor: 가위바위보 상성관계 리팩토링

newJunsung committed:
refactor: 에러 enum 추가 및 예외처리

forestjae commented on RockPaperScissors/RockPaperScissors/RPCError.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+enum RPCError: Error {
```
RPC는 어떤걸 의미하나요?

newJunsung replied:
묵찌빠 게임을 하면서 발생하는 오류들을 담아놓은 열거형 타입입니다.
프로젝트를 진행하면서 해당 열거형에 에러만 담을 것 같지 않기 때문에
(ex. 사용자와 컴퓨터가 동일한 모양을 내는 행위는 에러는 아닌 것 같습니다.),
열거형의 이름을 PlayingGameException으로 바꿀 예정입니다!

forestjae commented on RockPaperScissors/RockPaperScissors/RockPaperScissors.swift:
```diff
@@ -0,0 +1,27 @@
+    case rock = 2
+    case paper = 3
+    
+    static func < (lhs: RockPaperScissors, rhs: RockPaperScissors) -> Bool {
+        if lhs == .paper && rhs == .scissors {
+            return true
+        }
+        if lhs == .rock && rhs == .paper {
+            return true
+        }
+        if lhs == .scissors && rhs == .rock {
+            return true
+        }
+        return false
+    }
```
Comparable 프로토콜을 정의해서 가위바위보의 승패를 판정하는 아이디어 좋습니다 👍 👍 👍

newJunsung replied:
감사합니다!

forestjae commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,40 @@
+var isContinue = true
+while isContinue {
+    print("가위(1), 바위(2), 보(3)! <종료: 0> : ", terminator: "")
+    let str = readLine() ?? "error"
```
swift에서는 str이나 num같은 축약형 네이밍을 지양합니다.

https://www.swift.org/documentation/api-design-guidelines/

위 문서는 애플이 공식으로 제공하는 가이드라인인데요, 이 중 naming 부분을 읽어보시기 바랍니다.

silverjaehyeok replied:
공식 가이드라인을 참고하여 str을 사용자의 입력받아 상수에 할당하는 코드이기 때문에 userInput으로  수정하였고, num이 아닌 playerHandType으로 축약형 네이밍을 직관적으로 수정하였습니다.

forestjae commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,40 @@
+    if playerHand == computerHand {
+        print("비겼습니다!")
+        return true
+    }else if computerHand < playerHand {
+        print("이겼습니다!")
+        return false
+    }else{
+        print("졌습니다!")
+        return false
+    }
```
} 뒤와 { 앞에 띄어쓰기가 빠졌네요. 기본적인 컨벤션이니 꼭 지켜주세요!

silverjaehyeok replied:
네 수정했습니다. PR 요청 전에 한번 더 작성한 코드를 꼼꼼히 확인하는 습관을 가지도록 하겠습니다!

forestjae commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,40 @@
+    }
+    
+    do {
+        isContinue = try battle(player: num)
```
0을 입력하면 어떻게 되나요?

silverjaehyeok replied:
0을 입력 받았을 때 종료되지 않았던 문제를 아래 코드를 추가해 해결하였습니다.
```Swift 

    if playerHandType == 0 {
        print("게임 종료")
        break
    }
```

forestjae reviewed (changes requested):
수고 많으셨습니다. 👍 

Step1 요구사항에 부합하지 않는 점이 있어 수정 부탁드려요!

forestjae commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,40 @@
+    
+    do {
+        isContinue = try battle(player: num)
+    } catch {
+        print("잘못된 입력입니다. 다시 시도해주세요.")
+    }
```
do catch 구문을 잘 활용해주셨네요!

silverjaehyeok committed:
style: 변수 및 열거형 이름 변경

newJunsung committed:
style: battle함수 개선

기존 battle 함수는 매개변수로 player 하나 만을 받기 때문에,
두 명이 플레이 할 경우를 만들 수가 없었습니다.
따라서 battle의 매개변수를 2개를 두어, 추후에 player를 추가해도 승부를 낼 수 있도록
했습니다.

silverjaehyeok committed:
fix: 0을 입력받아도 게임이 종료되지 않던 버그 수정
